Add Crossfire and Deliantra multiplayer RPG servers

ToxicFrog opened the pull request:
###### Motivation for this change

This adds the necessary packages and NixOS configuration options to self-host a Crossfire or Deliantra server (a fork of an older version of Crossfire, with changes focusing on UX and new-player experience). Or both! Crossfire -- which saw its last official release five years ago, but has been under constant development since -- gets both a `-stable` version for the latest release and a `-latest` version for the SVN build.

It also includes version 1.73.0 of the Crossfire GTK client.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

ToxicFrog committed:
crossfire-server: init at r22111

This adds the Crossfire server package and the corresponding data packages
(maps and archetypes).

Note that this tracks a recent SVN commit; the latest official release is
many years old and no longer widely used (or supported).

ToxicFrog committed:
blitz: reinstate at 1.0.1

ToxicFrog committed:
nixos/deliantra-server: add settings module for Deliantra MMORPG server

ToxicFrog committed:
nixos/crossfire-server: add settings module for Crossfire MMORPG server

ToxicFrog committed:
perlPackages.DigestMD5: init at 2.55

ToxicFrog committed:
perlPackages.BDB: init at 1.92

ToxicFrog committed:
perlPackages.CompressLZF: init at 3.8

ToxicFrog committed:
perlPackages.ObjectEvent: init at 1.23

ToxicFrog committed:
perlPackages.AnyEventBDB: init at 1.1

ToxicFrog committed:
perlPackages.AnyEventIRC: init at 0.97

ToxicFrog committed:
perlPackages.CoroEV: init at 6.55

ToxicFrog committed:
perlPackages.Deliantra: init at 2.01

ToxicFrog committed:
deliantra-server: init at 3.1

Includes deliantra-arch, deliantra-maps, and deliantra-data (a combined
package that includes the compiled archetypes and maps and the static
data shipped with the server itself).

ToxicFrog committed:
crossfire-client: init at 1.75.0

The client is still getting official releases on a fairly frequent basis, so
this package tracks those rather than SVN head.

ToxicFrog commented:
Updated. Since this PR was opened one of the dependencies was deleted (the commit updating it was replaced with one re-adding it at a more recent version) and the default perl version changed.

ToxicFrog commented:
Ok, I think this is finally ready for review (again).

Mic92 commented on pkgs/games/crossfire/crossfire-arch.nix:
```diff
@@ -0,0 +1,26 @@
+
+  meta = with stdenv.lib; {
+    description = "Archetype data for the Crossfire free MMORPG";
+    homepage = http://crossfire.real-time.com/;
```
```suggestion
    homepage = "http://crossfire.real-time.com/";
```

Mic92 commented on pkgs/games/crossfire/crossfire-client.nix:
```diff
@@ -0,0 +1,25 @@
+
+  meta = with stdenv.lib; {
+    description = "GTKv2 client for the Crossfire free MMORPG";
+    homepage = http://crossfire.real-time.com/;
```
```suggestion
    homepage = "http://crossfire.real-time.com/";
```

Mic92 commented on pkgs/games/crossfire/crossfire-maps.nix:
```diff
@@ -0,0 +1,26 @@
+
+  meta = with stdenv.lib; {
+    description = "Map data for the Crossfire free MMORPG";
+    homepage = http://crossfire.real-time.com/;
```
```suggestion
    homepage = "http://crossfire.real-time.com/";
```

Mic92 commented on pkgs/games/crossfire/crossfire-server.nix:
```diff
@@ -0,0 +1,34 @@
+
+  meta = with stdenv.lib; {
+    description = "Server for the Crossfire free MMORPG";
+    homepage = http://crossfire.real-time.com/;
```
```suggestion
    homepage = "http://crossfire.real-time.com/";
```

Mic92 commented on pkgs/games/deliantra/deliantra-arch.nix:
```diff
@@ -0,0 +1,26 @@
+
+  meta = with stdenv.lib; {
+    description = "Archetype data for the Deliantra free MMORPG";
+    homepage = http://www.deliantra.net/;
```
```suggestion
    homepage = "http://www.deliantra.net/";
```

Mic92 commented on pkgs/games/deliantra/deliantra-data.nix:
```diff
@@ -0,0 +1,16 @@
+
+  meta = with stdenv.lib; {
+    description = "Combined game data (maps + archetypes) for the Deliantra free MMORPG";
+    homepage = http://www.deliantra.net/;
```
```suggestion
    homepage = "http://www.deliantra.net/";
```

Mic92 commented on pkgs/games/deliantra/deliantra-maps.nix:
```diff
@@ -0,0 +1,26 @@
+
+  meta = with stdenv.lib; {
+    description = "Map data for the Deliantra free MMORPG";
+    homepage = http://www.deliantra.net/;
```
```suggestion
    homepage = "http://www.deliantra.net/";
```

Mic92 commented on pkgs/games/deliantra/deliantra-server.nix:
```diff
@@ -0,0 +1,39 @@
+
+  meta = with stdenv.lib; {
+    description = "Server for the Deliantra free MMORPG";
+    homepage = http://www.deliantra.net/;
```
```suggestion
    homepage = "http://www.deliantra.net/";
```

Mic92 commented on pkgs/games/deliantra/deliantra-data.nix:
```diff
@@ -0,0 +1,16 @@
+    description = "Combined game data (maps + archetypes) for the Deliantra free MMORPG";
+    homepage = http://www.deliantra.net/;
+    license = licenses.gpl2;
+    platforms = platforms.linux;
```
Since it is just a symlink, we don't need to build this on hydra.

Mic92 commented on pkgs/games/crossfire/crossfire-server.nix:
```diff
@@ -0,0 +1,34 @@
+    rev = rev;
+  };
+
+  nativeBuildInputs = with pkgs; [autoconf automake libtool flex perl check pkg-config];
```
```suggestion
  nativeBuildInputs = [autoconf automake libtool flex perl check pkg-config];
```

ToxicFrog replied:
Breaks the build.

Mic92 replied:
You need to add all packages in the list to the header as well.

Mic92 commented on pkgs/games/crossfire/crossfire-client.nix:
```diff
@@ -0,0 +1,25 @@
+    rev = rev;
+  };
+
+  nativeBuildInputs = with pkgs; [cmake pkg-config perl vala];
```
```suggestion
  nativeBuildInputs = [cmake pkg-config perl vala];
```

ToxicFrog replied:
Breaks the build.

Mic92 replied:
see above.

Mic92 commented on pkgs/games/crossfire/crossfire-client.nix:
```diff
@@ -0,0 +1,25 @@
+  };
+
+  nativeBuildInputs = with pkgs; [cmake pkg-config perl vala];
+  buildInputs = with pkgs; [gtk2 pcre zlib libpng fribidi harfbuzzFull xorg.libpthreadstubs
```
```suggestion
  buildInputs = [gtk2 pcre zlib libpng fribidi harfbuzzFull xorg.libpthreadstubs
```

ToxicFrog replied:
Breaks the build.

Mic92 replied:
see above.

Mic92 commented on pkgs/development/libraries/blitz/default.nix:
```diff
@@ -0,0 +1,75 @@
+  };
+
+  nativeBuildInputs = [ pkgconfig python ];
+  buildInputs = [ gfortran texinfo ]
```
I think texinfo also belongs to `nativeBuildInputs`.

ToxicFrog replied:
I will take your word for it.

Mic92 commented on nixos/modules/services/games/deliantra-server.nix:
```diff
@@ -0,0 +1,153 @@
+
+    dataDir = mkOption {
+      type = types.str;
+      default = "${pkgs.deliantra-data}";
```
Also add `defaultText` here.

Mic92 commented on nixos/modules/services/games/crossfire-server.nix:
```diff
@@ -0,0 +1,119 @@
+      description = ''
+        Text to append to the corresponding configuration files. For the format
+        of each of these, consult the documentation in the corresponding file
+        in ${cfg.package}/etc/.
```
Avoid `${cfg.package}` here to avoid manual rebuilds if the package is changes.

ToxicFrog replied:
What would you recommend instead? If I point them at `/etc/crossfire` they don't have the docs  until the package is enabled; there's no good online documentation to point them at at; and it seems a bit mean to tell them to rummage around "somewhere in `/nix/store`" for it.

Mic92 replied:
Maybe online documentation or something like: 

```
cd $(nix-build '<nixpkgs>' -A crossfire --no-out-link)/etc
```

Mic92 commented on nixos/modules/services/games/crossfire-server.nix:
```diff
@@ -0,0 +1,119 @@
+
+    dataDir = mkOption {
+      type = types.str;
+      default = "${cfg.package}/share/crossfire";
```
`defaultText`.

Mic92 commented on pkgs/games/crossfire/default.nix:
```diff
@@ -0,0 +1,38 @@
+  crossfire-client = callPackage ./crossfire-client.nix {
+    version = "1.73.0"; rev = 20680;
+    sha256 = "192y9nmskrx907h16p8rl4cvksnh5m0q8kz9cxw376sq7gam7g0y";
+  };
```
Can we limit the number of versions here?

ToxicFrog replied:
How do you mean? Ditch the latest/stable separation? Enough has changed (with no official release) since 1.7.1 that I think it does make sense to have both available -- there are still 171 servers out there.

Mic92 commented on nixos/modules/services/games/crossfire-server.nix:
```diff
@@ -0,0 +1,120 @@
+    dataDir = mkOption {
+      type = types.str;
+      default = "${cfg.package}/share/crossfire";
+      defaultText = "\${services.crossfire.package}/share/crossfire";
```
```suggestion
      defaultText = "\${config.services.crossfire.package}/share/crossfire";
```

Mic92 commented on pkgs/games/crossfire/crossfire-arch.nix:
```diff
@@ -0,0 +1,26 @@
+
+  phases = ["unpackPhase" "installPhase"];
+  installPhase = ''
+    mkdir -p "$out"
```
This is a no-op. You can just add a meta attribute to `fetchsvn` and return it instead.

ToxicFrog replied:
I'm not sure you mean by this.

Mic92 replied:
Right now you are building a derivation that returns the content of the resource attribute.
This creates an identical derivation with identical content of the `fetchsvn` derivation. 
Instead you can just say:

```
fetchsvn {
    url = "http://svn.code.sf.net/p/crossfire/code/arch/trunk/";
    sha256 = sha256;
    rev = rev;
    meta = with lib; {
      description = "Archetype data for the Crossfire free MMORPG";
      homepage = "http://crossfire.real-time.com/";
      license = licenses.gpl2;
      platforms = platforms.linux;
      maintainers = with maintainers; [ ToxicFrog ];
  };
};
```

instead of 
```
stdenv.mkDerivation {
   ...
   src = ...;
}
```

Mic92 commented on pkgs/games/crossfire/crossfire-maps.nix:
```diff
@@ -0,0 +1,26 @@
+  name = "crossfire-maps-${version}";
+
+  src = fetchsvn {
+    url = "http://svn.code.sf.net/p/crossfire/code/maps/trunk/";
```
same as above.

Mic92 commented on pkgs/games/deliantra/deliantra-arch.nix:
```diff
@@ -0,0 +1,26 @@
+  installPhase = ''
+    mkdir -p "$out"
+    export DELIANTRA_DATADIR="$out"
+    ${deliantra-server}/bin/cfutil --install-arch .
```
If it is not very CPU intensive we could just set `hydraPlatforms = []` to build this on the clients itself.

ToxicFrog replied:
Done.

Mic92 commented on pkgs/games/deliantra/deliantra-maps.nix:
```diff
@@ -0,0 +1,26 @@
+  installPhase = ''
+    mkdir -p "$out/maps"
+    export DELIANTRA_DATADIR="$out"
+    ${deliantra-server}/bin/cfutil --install-maps .
```
same as above.

ToxicFrog replied:
Done.

thiagokokada reviewed (commented):
Please fix the merge conflict.

Also, could you change the PR title? I think it is really non-descriptive as-is and I was wondering what this PR was about.

ToxicFrog commented:
I think this is ready now.

ToxicFrog commented:
Beep?

Mic92 commented on pkgs/games/deliantra/deliantra-data.nix:
```diff
@@ -0,0 +1,17 @@
+  name = "deliantra-data-${version}";
+  version = "M${deliantra-maps.version}+A${deliantra-arch.version}";
+
+  hydraPlatforms = [];
```
hydraPlatforms needs to be in meta to have an effect.

Mic92 commented on pkgs/games/deliantra/deliantra-maps.nix:
```diff
@@ -0,0 +1,27 @@
+    sha256 = "0zbwzya28s1xpnbrmqkqvfrzns03zdjd8a9w9nk665aif6rw2zbz";
+  };
+
+  hydraPlatforms = [];
```
same here.

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-server.nix:
```diff
@@ -0,0 +1,41 @@
+  meta = with lib; {
+    description = "Server for the Deliantra free MMORPG";
+    homepage = "http://www.deliantra.net/";
+    license = licenses.gpl2;
```
Do you know if the license is gpl2Only or gpl2Plus?

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-server.nix:
```diff
@@ -0,0 +1,41 @@
+  propagatedBuildInputs = perl-deps;
+
+  hardeningDisable = [ "format" ];
+  patches = [ ./0001-abs.patch ./0002-datadir.patch ];
```
```suggestion
  patches = [
    ./0001-abs.patch
    ./0002-datadir.patch
  ];
```

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-server.nix:
```diff
@@ -0,0 +1,41 @@
+    sha256 = "0v0m2m9fxq143aknh7jb3qj8bnpjrs3bpbbx07c18516y3izr71d";
+  };
+
+  nativeBuildInputs = [
+    autoconf perl gperf optipng pngnq rsync imagemagick blitz
+    pkg-config glib boost makeWrapper
+  ];
```
Some of those are buildInputs.

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-server.nix:
```diff
@@ -0,0 +1,41 @@
+    Deliantra DigestSHA1 EV PodPOM SafeHole URI YAMLLibYAML
+  ];
+in
+stdenv.mkDerivation rec {
+  name = "deliantra-server-${version}";
+  version = "3.1";
+
+  src = fetchurl {
+    url = "http://dist.schmorp.de/deliantra/deliantra-server-3.1.tar.xz";
```
```suggestion
stdenv.mkDerivation rec {
  pname = "deliantra-server";
  version = "3.1";

  src = fetchurl {
    url = "http://dist.schmorp.de/deliantra/deliantra-server-${version}.tar.xz";
```

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-server.nix:
```diff
@@ -0,0 +1,41 @@
+{ stdenv, lib, fetchurl, perlPackages,
+  autoconf, perl, gperf, optipng, pngnq, rsync, imagemagick, blitz,
+  pkg-config, glib, boost, makeWrapper }:
```
```suggestion
{ stdenv, lib, fetchurl, perlPackages
, autoconf, perl, gperf, optipng, pngnq, rsync, imagemagick, blitz
, pkg-config, glib, boost, makeWrapper
}:
```

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-maps.nix:
```diff
@@ -0,0 +1,27 @@
+  meta = with lib; {
+    description = "Map data for the Deliantra free MMORPG";
+    homepage = "http://www.deliantra.net/";
+    license = licenses.gpl2;
```
Do you know if the license is gpl2Only or gpl2Plus?

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-maps.nix:
```diff
@@ -0,0 +1,27 @@
+  };
+
+  hydraPlatforms = [];
+  phases = ["unpackPhase" "installPhase"];
```
```suggestion

```

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-maps.nix:
```diff
@@ -0,0 +1,27 @@
+{ stdenv, lib, fetchurl, deliantra-server }:
+
+stdenv.mkDerivation rec {
+  name = "deliantra-maps-${version}";
+  version = "3.1";
+
+  src = fetchurl {
+    url = "http://dist.schmorp.de/deliantra/${name}.tar.xz";
```
```suggestion
  pname = "deliantra-maps";
  version = "3.1";

  src = fetchurl {
    url = "http://dist.schmorp.de/deliantra/deliantra-maps-${version}.tar.xz";
```

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-data.nix:
```diff
@@ -0,0 +1,17 @@
+  meta = with lib; {
+    description = "Combined game data (maps + archetypes) for the Deliantra free MMORPG";
+    homepage = "http://www.deliantra.net/";
+    license = licenses.gpl2;
```
Do you know if the license is gpl2Only or gpl2Plus?

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-data.nix:
```diff
@@ -0,0 +1,17 @@
+{ stdenv, lib, fetchurl, deliantra-maps, deliantra-arch, deliantra-server, symlinkJoin }:
+
+symlinkJoin rec {
+  name = "deliantra-data-${version}";
+  version = "M${deliantra-maps.version}+A${deliantra-arch.version}";
+
```
```suggestion
  pname = "deliantra-data";
  version = "M${deliantra-maps.version}+A${deliantra-arch.version}";

```

ToxicFrog replied:
symlinkJoin doesn't support pname.

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-arch.nix:
```diff
@@ -0,0 +1,27 @@
+  meta = with lib; {
+    description = "Archetype data for the Deliantra free MMORPG";
+    homepage = "http://www.deliantra.net/";
+    license = licenses.gpl2;
```
Do you know if the license is gpl2Only or gpl2Plus?

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-arch.nix:
```diff
@@ -0,0 +1,27 @@
+  };
+
+  hydraPlatforms = [];
+  phases = ["unpackPhase" "installPhase"];
```
```suggestion

```

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-arch.nix:
```diff
@@ -0,0 +1,27 @@
+    sha256 = "1xzhv48g90hwkzgx9nfjm81ivg6hchkik9ldimi8ijb4j393kvsz";
+  };
+
+  hydraPlatforms = [];
```
needs to be in meta.

SuperSandro2000 commented on pkgs/games/deliantra/deliantra-arch.nix:
```diff
@@ -0,0 +1,27 @@
+{ stdenv, lib, fetchurl, deliantra-server }:
+
+stdenv.mkDerivation rec {
+  name = "deliantra-arch-${version}";
+  version = "3.1";
+
+  src = fetchurl {
+    url = "http://dist.schmorp.de/deliantra/${name}.tar.xz";
```
```suggestion
  pname = "deliantra-arch";
  version = "3.1";

  src = fetchurl {
    url = "http://dist.schmorp.de/deliantra/deliantra-arch-${version}.tar.xz";
```

SuperSandro2000 commented on pkgs/games/deliantra/default.nix:
```diff
@@ -0,0 +1,10 @@
+pkgs:
+
+let
+  callPackage = pkgs.callPackage;
+in {
+  deliantra-server = callPackage ./deliantra-server.nix {};
+  deliantra-arch = callPackage ./deliantra-arch.nix {};
+  deliantra-maps = callPackage ./deliantra-maps.nix {};
+  deliantra-data = callPackage ./deliantra-data.nix {};
+}
```
All this could also be in top-level.

ToxicFrog replied:
It could be, but top-level is already so massive that I'd rather push the boilerplate down here.

SuperSandro2000 commented on pkgs/games/crossfire/default.nix:
```diff
@@ -0,0 +1,38 @@
+  callPackage = pkgs.callPackage;
+in rec {
+  crossfire-client = callPackage ./crossfire-client.nix {
+    version = "1.73.0"; rev = 20680;
```
```suggestion
    version = "1.73.0";
    rev = 20680;
```

SuperSandro2000 commented on pkgs/games/crossfire/default.nix:
```diff
@@ -0,0 +1,38 @@
+  };
+
+  crossfire-server-stable = callPackage ./crossfire-server.nix {
+    version = "1.71.0"; rev = 19312;
+    sha256 = "0ymr31n9w88yd7nb3lfgiwjppvsz77ajqdg47l2b9d2fl342sm9i";
+    maps = crossfire-maps-stable; arch = crossfire-arch-stable;
+  };
+  crossfire-arch-stable = callPackage ./crossfire-arch.nix {
+    version = "1.71.0"; rev = 19308;
+    sha256 = "1jl4a6yy0hvf416nyhga7mq4wns6c6w91aqd52p2mh1zny428c3n";
+  };
+  crossfire-maps-stable = callPackage ./crossfire-maps.nix {
+    version = "1.71.0"; rev = 19309;
+    sha256 = "1rplbrac5dyw742zijfdq63s7ij7d5277fscas0a5p50768ymyrp";
+  };
+
+  crossfire-server-latest = callPackage ./crossfire-server.nix {
+    version = "latest"; rev = 20926;
+    sha256 = "00z8c9bk6gkyggl47klw1gadlzid5cimallz184cppsxbngqmsx6";
+    maps = crossfire-maps-latest; arch = crossfire-arch-latest;
+  };
+  crossfire-arch-latest = callPackage ./crossfire-arch.nix {
+    version = "latest"; rev = 20926;
+    sha256 = "12yc9cn9yxazaagh2csi9djph6mfcggxvaxkra6fd10x2nbczw4g";
+  };
+  crossfire-maps-latest = callPackage ./crossfire-maps.nix {
+    version = "latest"; rev = 20926;
```
```suggestion
    version = "1.71.0";
    rev = 19312;
    sha256 = "0ymr31n9w88yd7nb3lfgiwjppvsz77ajqdg47l2b9d2fl342sm9i";
    maps = crossfire-maps-stable; arch = crossfire-arch-stable;
  };
  crossfire-arch-stable = callPackage ./crossfire-arch.nix {
    version = "1.71.0";
    rev = 19308;
    sha256 = "1jl4a6yy0hvf416nyhga7mq4wns6c6w91aqd52p2mh1zny428c3n";
  };
  crossfire-maps-stable = callPackage ./crossfire-maps.nix {
    version = "1.71.0";
    rev = 19309;
    sha256 = "1rplbrac5dyw742zijfdq63s7ij7d5277fscas0a5p50768ymyrp";
  };

  crossfire-server-latest = callPackage ./crossfire-server.nix {
    version = "latest";
    rev = 20926;
    sha256 = "00z8c9bk6gkyggl47klw1gadlzid5cimallz184cppsxbngqmsx6";
    maps = crossfire-maps-latest; arch = crossfire-arch-latest;
  };
  crossfire-arch-latest = callPackage ./crossfire-arch.nix {
    version = "latest";
    rev = 20926;
    sha256 = "12yc9cn9yxazaagh2csi9djph6mfcggxvaxkra6fd10x2nbczw4g";
  };
  crossfire-maps-latest = callPackage ./crossfire-maps.nix {
    version = "latest";
    rev = 20926;
```

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-server.nix:
```diff
@@ -0,0 +1,34 @@
+  meta = with lib; {
+    description = "Server for the Crossfire free MMORPG";
+    homepage = "http://crossfire.real-time.com/";
+    license = licenses.gpl2;
```
Do you know if the license is gpl2Only or gpl2Plus?

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-server.nix:
```diff
@@ -0,0 +1,34 @@
+    rev = rev;
+  };
+
+  nativeBuildInputs = [autoconf automake libtool flex perl check pkg-config python2];
```
```suggestion
  nativeBuildInputs = [ autoconf automake libtool flex perl check pkg-config python2 ];
```
can we use python3?

ToxicFrog replied:
At the time I wrote this, python3 support was still in the early planning stages. It looks like may be at least partially supported now, but I'm hesitant to change too much considering how long this has been in flight; I'd rather commit it in its known-working state and then do a followup PR that updates crossfire-server-latest to 2021 and, if possible, updates the python dep to py3.

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-server.nix:
```diff
@@ -0,0 +1,34 @@
+  version, rev, sha256, maps, arch }:
+
+stdenv.mkDerivation rec {
+  version = "r${toString rev}";
+  name = "crossfire-server-${version}";
```
```suggestion
  pname = "crossfire-server";
  version = "r${rev}";
```

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-maps.nix:
```diff
@@ -0,0 +1,28 @@
+  meta = with lib; {
+    description = "Map data for the Crossfire free MMORPG";
+    homepage = "http://crossfire.real-time.com/";
+    license = licenses.gpl2;
```
Do you know if the license is gpl2Only or gpl2Plus?

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-maps.nix:
```diff
@@ -0,0 +1,28 @@
+  };
+
+  hydraPlatforms = [];
+  phases = ["unpackPhase" "installPhase"];
```
```suggestion

```

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-maps.nix:
```diff
@@ -0,0 +1,28 @@
+  version, rev, sha256 }:
+
+stdenv.mkDerivation rec {
+  version = "r${toString rev}";
+  name = "crossfire-maps-${version}";
```
```suggestion
  pname = "crossfire-maps";
  version = "r${rev}";
```

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-client.nix:
```diff
@@ -0,0 +1,29 @@
+  meta = with lib; {
+    description = "GTKv2 client for the Crossfire free MMORPG";
+    homepage = "http://crossfire.real-time.com/";
+    license = licenses.gpl2;
```
Do you know if the license is gpl2Only or gpl2Plus?

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-client.nix:
```diff
@@ -0,0 +1,29 @@
+    rev = rev;
+  };
+
+  nativeBuildInputs = [cmake pkg-config perl vala];
+  buildInputs = [gtk2 pcre zlib libpng fribidi harfbuzzFull xorg.libpthreadstubs
+    xorg.libXdmcp curl SDL SDL_image SDL_mixer util-linux libselinux libsepol];
```
```suggestion
  nativeBuildInputs = [ cmake pkg-config perl vala ];
  buildInputs = [ gtk2 pcre zlib libpng fribidi harfbuzzFull xorg.libpthreadstubs
    xorg.libXdmcp curl SDL SDL_image SDL_mixer util-linux libselinux libsepol ];
```

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-client.nix:
```diff
@@ -0,0 +1,29 @@
+  version, rev, sha256 }:
+
+stdenv.mkDerivation rec {
+  version = "r${toString rev}";
+  name = "crossfire-client-${version}";
```
```suggestion
  pname = "crossfire-client";
  version = "r${rev}";
```

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-client.nix:
```diff
@@ -0,0 +1,29 @@
+{ stdenv, lib, fetchsvn,
+  cmake, pkg-config, perl, vala,
+  gtk2, pcre, zlib, libpng, fribidi, harfbuzzFull, xorg, util-linux, curl,
+  SDL, SDL_image, SDL_mixer, libselinux, libsepol,
+  version, rev, sha256 }:
```
```suggestion
{ stdenv, lib, fetchsvn
, cmake, pkg-config, perl, vala
, gtk2, pcre, zlib, libpng, fribidi, harfbuzzFull, xorg, util-linux, curl
, SDL, SDL_image, SDL_mixer, libselinux, libsepol
, version, rev, sha256
}:
```

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-arch.nix:
```diff
@@ -0,0 +1,28 @@
+  meta = with lib; {
+    description = "Archetype data for the Crossfire free MMORPG";
+    homepage = "http://crossfire.real-time.com/";
+    license = licenses.gpl2;
```
Do you know if the license is gpl2Only or gpl2Plus?

ToxicFrog replied:
Looks like gpl2Plus, based on the source headers.

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-arch.nix:
```diff
@@ -0,0 +1,28 @@
+  };
+
+  hydraPlatforms = [];
+  phases = ["unpackPhase" "installPhase"];
```
```suggestion

```

SuperSandro2000 commented on pkgs/games/crossfire/crossfire-arch.nix:
```diff
@@ -0,0 +1,28 @@
+  version, rev, sha256 }:
+
+stdenv.mkDerivation rec {
+  version = "r${toString rev}";
+  name = "crossfire-arch-${version}";
```
```suggestion
  pname = "crossfire-arch";
  version = "r${rev}";
```

SuperSandro2000 commented on pkgs/development/libraries/blitz/default.nix:
```diff
@@ -0,0 +1,75 @@
+  meta = with lib; {
+    description = "Fast multi-dimensional array library for C++";
+    homepage = https://sourceforge.net/projects/blitz/;
+    license = licenses.lgpl3;
+    platforms = platforms.linux ++ platforms.darwin;
+    maintainers = with maintainers; [ ToxicFrog ];
+
+    longDescription = ''
```
```suggestion
    license = licenses.lgpl3;
    platforms = platforms.unix;
    maintainers = with maintainers; [ ToxicFrog ];
    longDescription = ''
```

SuperSandro2000 commented on pkgs/development/libraries/blitz/default.nix:
```diff
@@ -0,0 +1,75 @@
+    ]
+    ++ optional enablePadding "--enable-array-length-padding"
+    ++ optional enableSerialization "--enable-serialization"
+    ++ optionals (boost != null) [ "--with-boost=${boost.dev}"
+                                   "--with-boost-libdir=${boost.out}/lib" ]
```
```suggestion
    ++ optionals (boost != null) [
      "--with-boost=${boost.dev}"
      "--with-boost-libdir=${boost.out}/lib"
    ]
```

SuperSandro2000 commented on pkgs/development/libraries/blitz/default.nix:
```diff
@@ -0,0 +1,75 @@
+
+}:
+
+assert enableSerialization -> boost != null;
+
+let
+  inherit (lib) optional optionals;
+in
+
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation rec {
```

ToxicFrog replied:
I have left some of this intact, since optional and optionals are both needed and boost is only required if serialization is enabled.

SuperSandro2000 commented on pkgs/development/libraries/blitz/default.nix:
```diff
@@ -0,0 +1,75 @@
+in
+
+stdenv.mkDerivation rec {
+  name = "blitz++-1.0.1";
+  src = fetchurl {
+    url = "https://github.com/blitzpp/blitz/archive/1.0.1.tar.gz";
+    sha256 = "1vyw795jxg7d9ac7hlybink0jgkrybjcbzh1gcq69ck4ggqc6bxn";
```
```suggestion
  pname = "blitz++";
  version = "1.0.1";

  src = fetchFromGitHub{
    ...
```

SuperSandro2000 commented on pkgs/development/libraries/blitz/default.nix:
```diff
@@ -0,0 +1,75 @@
+{ stdenv, lib, fetchurl, pkg-config, gfortran, texinfo, python
+
+# Select SIMD alignment width (in bytes) for vectorization.
+, simdWidth ? 1
+
+# Pad arrays to simdWidth by default?
+# Note: Only useful if simdWidth > 1
+, enablePadding ? false
+
+# Activate serialization through Boost.Serialize?
+, enableSerialization ? true, boost ? null
+
+# Activate test-suite?
+# WARNING: Some of the tests require up to 1700MB of memory to compile.
+, doCheck ? true
+
+}:
```
```suggestion
{ stdenv, lib, fetchurl, pkg-config, gfortran, texinfo, python
# Select SIMD alignment width (in bytes) for vectorization.
, simdWidth ? 1
# Pad arrays to simdWidth by default?
# Note: Only useful if simdWidth > 1
, enablePadding ? false
, enableSerialization ? true, boost
# WARNING: Some of the tests require up to 1700MB of memory to compile.
, doCheck ? true
}:
```

ToxicFrog commented:
All review comments addressed; a few weren't applied, or were applied only with modifications:
- `fetchsvn` cannot take a `meta` attribute
- `"${foo}"` is not equivalent to `"${toString foo}"` when foo is an integer
- `blitz`'s usage of `optional`/`optionals` is necessary

And a few other problems were found and addressed:
- `<` and `>` need to be escaped in option documentation or `docbook` falls over
- `deliantra-arch` and `deliantra-maps` are (unlike the `crossfire` versions, which can be used as is) fairly slow to build, so I removed `hydraPlatforms = []` from them

ToxicFrog commented:
Also, github says I have a change requested by @SuperSandro2000 , but it also won't show me what the change actually _is_, and the "changed files" tab lists all conversations as resolved, so...hopefully I dealt with whatever that was along with everything else.

aanderse reviewed (changes requested):
Hello! I have left some comments that I hope are helpful. If you want any help on this I'd be happy to chip in. I think this PR is very close to a merge... let's push it past the finish line! :rocket:

ToxicFrog commented on nixos/modules/services/games/crossfire-server.nix:
```diff
@@ -0,0 +1,121 @@
+    users.users.crossfire = {
+      description     = "Crossfire server daemon user";
+      home            = cfg.runtimeDir;
+      createHome      = true;
```
Done. Added a warning about the directory not being auto-created if you change the default to the `stateDir` option.

ToxicFrog commented:
Alright, I think this is ready, or will be once ofborg finishes.

aanderse commented on nixos/modules/services/games/crossfire-server.nix:
```diff
@@ -0,0 +1,121 @@
+
+      preStart = ''
+        if [ ! -e ${cfg.runtimeDir}/bookarch ]; then
+          ${pkgs.rsync}/bin/rsync -a --chmod=u=rwX,go=rX \
```
Perfect! Thanks for explanation.

aanderse commented on nixos/modules/services/games/crossfire-server.nix:
```diff
@@ -0,0 +1,177 @@
+      description = ''
+        Where to store runtime data (save files, persistent items, etc).
+
+        If left at the default, this will be automatically created the first
```
It's actually **every** time the server starts.

aanderse commented on pkgs/games/deliantra/deliantra-server.nix:
```diff
@@ -0,0 +1,48 @@
+}:
+
+let
+  perl-deps = with perlPackages; [
```
This is totally fine but I thought I would let you know about `perl.withPackages` which probably works here as well, just in case you didn't know about it.

ToxicFrog replied:
Thanks for the pointer! I didn't, but I think this looks cleaner for this application.

aanderse commented:
I love the modules! Such fantastic work. All the descriptions are clear enough that users won't have to read the module source code 👍 and the module source code is very well documented so that anyone who reads it will know exactly what and why things are the way they are 🚀 

Full bonus points awarded, and then some more! :sparkles:

ToxicFrog commented:
Updated the documentation for `stateDir` (on both crossfire and deliantra) to clarify directory autocreation behaviour. Thanks for the correction, and for all the work reviewing this more generally :)

aanderse reviewed (approved):
Modules LGTM :+1:

ToxicFrog commented:
Something about the last round of changes I made to address review comments on the modules have resulted in `nixos-rebuild` with the crossfire server enabled consuming large (20+GB) amounts of memory, so I'm going to look into that and report back when it's fixed.

ToxicFrog commented:
False alarm; my old overlays from before I added crossfire/deliantra to nixpkgs were interacting badly with the version in my local nixpkgs checkout.
